### PR TITLE
 Add partial display update support.

### LINKED ISSUE #1169: Support partial display updates
Every object within displayio groups should be able to provide rectangles to redraw for them.

### LINKED ISSUE #1705: displayio.Group rotation and flipping
It would be nice to be able to rotate the Group in 90° increments, and to be able to mirror and vertically flip it. One way of implementing it would be adding an ``orientation`` attribute, with eight possible values. One bit would control horizontal flip, one bit — vertical, and last bit would control swapping of the ``x`` and ``y`` coordinates (diagonal flip). This way we can have all possible orientations.

### LINKED ISSUE #1923: Displayio / MiniTFTFeatherWing noise at left edge of screen
The following code:
```
     import time
     from adafruit_featherwing import minitft_featherwing
     minitft = minitft_featherwing.MiniTFTFeatherWing()
     print('hello')
     while True:
         pass
```
Produces the following image with noise on the left side:
TFT_label_bug_2

This noise persists even if the display is more substantially over-written as in:

```
import time
import board
from adafruit_featherwing import minitft_featherwing
import displayio
import terminalio
from adafruit_display_text import label

minitft = minitft_featherwing.MiniTFTFeatherWing()

label_count = 5
labels = list()
group = displayio.Group(max_size=label_count, scale=1, x=0, y=0)
font = terminalio.FONT
f_width, f_height = font.get_bounding_box()

colors = [0xffffff, 0xb0b0b0, 0xffff00, 0xb0b0b0, 0x00d0d0]
for i in range(label_count):
    labels.append(label.Label(font, max_glyphs=37, color=colors[i], line_spacing=0.94))
    labels[-1].x = 5
    labels[-1].y = i * f_height + int(f_height*.5)
    labels[-1].text = 'line %d' % i
    group.append(labels[-1])

minitft.display.show(group)
while True:
    pass
```

Which produces:

TFT_label_bug

### PR DESCRIPTION
**Pull request by tannewt**
Different operations to the display tree have different costs. Be
aware of these costs when optimizing your code.
* Changing tiles indices in a TileGrid will update an area
covering them all.
* Changing a palette will refresh every object that references it.
* Moving a TileGrid will update both where it was and where it moved to.
* Adding something to a Group will refresh each individual area it
covers.
* Removing things from a Group will refresh one area that covers all
previous locations. (Not separate areas like add.)
* Setting a new top level Group will refresh the entire display.

Only TileGrid moves are optimized for overlap. All other overlaps
cause sending of duplicate pixels.

This also adds flip_x, flip_y and transpose_xy to TileGrid. They
change the direction of the pixels but not the location.

Fixes #1169. Fixes #1705. Fixes #1923.

### CODE REVIEW

File: shared-module/displayio/__init__.c [hunk 1]
```diff
@@ -19,14 +20,83 @@
...
+        }
+        remaining_rows -= rows_per_buffer;
+
+        displayio_display_begin_transaction(display);
+        displayio_display_set_region_to_update(display, subrectangle.x1, subrectangle.y1,
+                                                        subrectangle.x2, subrectangle.y2);
+        displayio_display_end_transaction(display);
+
+        uint32_t mask[(buffer_size / 32) + 1];
+        for (uint16_t k = 0; k < (buffer_size / 32) + 1; k++) {
```
Comment (dhalbert) [inline shared-module/displayio/__init__.c]: Would `memset()` be faster?

Comment (dhalbert) [reply shared-module/displayio/__init__.c]: maybe not - I looked on stackoverflow, and, it depends.

File: shared-module/displayio/Group.c [hunk 2]
```diff
@@ -38,60 +38,201 @@ uint32_t common_hal_displayio_group_get_scale(displayio_group_t* self) {
...
     return self->x;
 }
 
 void common_hal_displayio_group_set_x(displayio_group_t* self, mp_int_t x) {
-    self->needs_refresh = self->x != x;
+    if (self->x == x) {
+        return;
+    }
+    if (self->absolute_transform.transpose_xy) {
+        int8_t dy = self->absolute_transform.dy / self->scale;
```
[the remark this answers is not in the document]
Comment (tannewt) [reply shared-module/displayio/Group.c]: yup, good catch. in general coordinates are 16 bit and scale 8.

File: shared-module/displayio/TileGrid.c [hunk 3]
```diff
@@ -150,30 +297,40 @@ bool displayio_tilegrid_get_area(displayio_tilegrid_t *self, displayio_buffer_tr
...
+        flip_y = temp_flip;
     }
+
+    // How many pixels are outside of our area between us and the start of the row.
     uint16_t start = 0;
-    if (transform->mirror_x) {
-        start += (area->x2 - area->x1 - 1) * x_stride;
-        x_stride *= -1;
+    if ((self->absolute_transform->dx < 0) != flip_x) {
+        // if (self->absolute_transform->transpose_xy) {
```
[the remark this answers is not in the document]
Comment (tannewt) [reply shared-module/displayio/TileGrid.c]: was experimenting but not needed.

File: shared-module/displayio/__init__.c (hunk 1 above)
Comment (tannewt) [reply shared-module/displayio/__init__.c]: easier to read either way. :-)

Comment (tannewt) [reply shared-module/displayio/__init__.c]: Didn't work with naive memset. This has the advantage of knowing the array type and that it is aligned.

**Pull request merged**